p2p/protocol.go

```diff
@@ -0,0 +1,202 @@
+package p2p
+
+import (
+	"container/list"
+	"errors"
+	"github.com/gogo/protobuf/proto"
+	"github.com/spacemeshos/go-spacemesh/crypto"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p/pb"
+	"github.com/spacemeshos/go-spacemesh/p2p/service"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type MessageType uint32
+
+type Item struct {
+	id        uint64
+	timestamp time.Time
+}
+
+type Protocol struct {
+	id                 uint64
+	name               string
+	network            Service
+	pendMutex          sync.RWMutex
+	pendingMap         map[uint64]chan interface{}
+	pendingQueue       *list.List
+	resHandlers        map[uint64]func(msg []byte)
+	msgRequestHandlers map[MessageType]func(msg []byte) []byte
+	ingressChannel     chan service.Message
+	requestLifetime    time.Duration
+}
+
+func NewProtocol(network Service, name string, ttl time.Duration) *Protocol {
+	p := &Protocol{
+		name:               name,
+		pendingMap:         make(map[uint64]chan interface{}),
+		resHandlers:        make(map[uint64]func(msg []byte)),
+		pendingQueue:       list.New(),
+		network:            network,
+		ingressChannel:     network.RegisterProtocol(name),
+		msgRequestHandlers: make(map[MessageType]func(msg []byte) []byte),
+		requestLifetime:    ttl,
+	}
+
+	go p.readLoop()
+	return p
+}
+
+func (p *Protocol) readLoop() {
+	for {
+		timer := time.NewTimer(time.Second)
+		select {
+		case msg, ok := <-p.ingressChannel:
+			if !ok {
+				log.Error("read loop channel was closed")
+				break
+			}
+			//todo add buffer and option to limit number of concurrent goroutines
+			go p.handleMessage(msg)
+		case <-timer.C:
+			go p.cleanStaleMessages()
+		}
+	}
+}
+
+func (p *Protocol) cleanStaleMessages() {
+	for {
```

Comment:
you should run this under a writer lock

Reply:
Also, what about cleaning the resHandlers map?

Reply:
good idea, done.